### PR TITLE
feat(search): add cross-encoder reranker with position-aware score blending

### PR DESCRIPTION
**Pull request by aliasunder**
## Summary

- Add cross-encoder reranker (ms-marco-MiniLM-L-6-v2, q8) that refines hybrid search results after RRF fusion
- Position-aware score blending protects strong RRF hits at top ranks (75% retrieval / 25% reranker) while letting the reranker rescue demoted results at lower ranks (40% / 60%)
- New `RERANK_MODE` env var (default: `blended`, values: `none` | `blended`) flows through all deploy surfaces
- Response envelope gains `reranked: boolean` field so clients know when cross-encoder reranking was applied

## Details

**Phase 2 PR 4** — the final quality step for hybrid search. Fixes intent-heavy queries (e.g., "how I feel about AI tools") where the bi-encoder can't distinguish opinions from expertise. Spike-validated: 9/9 eval queries pass, 0 regressions, ~200ms added latency.

**New module `reranker.ts`**: factory-closure pattern (lazy-loading, promise guard — same as `embedder.ts`). Pure scoring functions (`blendScores`, `normalizeScores`) as named exports. Uses `AutoModelForSequenceClassification` + `AutoTokenizer` (not the `pipeline()` API) because text-classification pipelines don't support `text_pair` input.

**Blend weight tiers** (by 1-indexed RRF rank):
| Rank | RRF | Reranker | Rationale |
|------|-----|----------|-----------|
| 1–3  | 75% | 25%      | Protect strong retrieval hits |
| 4–10 | 50% | 50%      | Even blend in the middle |
| 11+  | 40% | 60%      | Let reranker rescue demoted results |

## Test plan

- [x] 25 new tests (20 unit + 5 integration) — `npm test` passes (1256 total)
- [x] `blendScores` table-driven tests covering all three tier boundaries + edge cases
- [x] `createReranker` factory tests with mocked `@huggingface/transformers`
- [x] Integration tests: `reranked: true/false` field, graceful fallback on reranker failure
- [x] Build + lint clean
- [x] Deploy surfaces: RERANK_MODE in all 6 docker-compose, 3 .env.example, cli/src/env.ts
- [x] Feature surface docs: README, ARCHITECTURE, AGENTS.md, DEPLOY.md, cli/README, server.json, wiki.json

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added optional cross-encoder reranking for hybrid search, with a new `RERANK_MODE` setting.
  * Search results now include a `reranked` flag so you can tell when reranking was applied.
* **Bug Fixes**
  * Hybrid search now falls back more gracefully when reranking isn’t available or fails, keeping results usable.
* **Documentation**
  * Updated setup, configuration, and search docs to reflect the new reranking behavior and options.

### CODE REVIEW

File: src/vault-mcp/search/search-queries.ts [hunk 1]
```diff
@@ -291,17 +297,96 @@ export const hybridSearch = async (
...
+          query: params.query,
+          mergedResults,
+          vectorHitsByPath,
+          selectFirstChunkStmt: context.selectFirstChunkStmt,
+          logger,
+        })
+      : null
+
+  const finalResults = rerankedResult?.results ?? mergedResults
+  const reranked = rerankedResult !== null
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply src/vault-mcp/search/search-queries.ts]: Fixed — capped the rerank window at `userLimit` (default 20) before passing to `tryRerank()`. Results beyond `userLimit` are sliced off after reranking anyway, so scoring them was wasted work. This matches the spike-validated behavior (20 candidates, ~200ms).

**Comment by aliasunder** [comment]
@coderabbit review

File: src/vault-mcp/search/search-queries.ts [hunk 2]
```diff
@@ -291,17 +295,98 @@ export const hybridSearch = async (
...
+
+    const rrfScores = params.mergedResults.map((result) => result.score)
+    const rrfRanks = params.mergedResults.map((_result, index) => index + 1)
+
+    const blendedScores = blendScores({ rrfScores, rerankScores, rrfRanks })
+
+    const scoredResults = params.mergedResults.map((result, index) => ({
+      ...result,
+      score: blendedScores[index],
+    }))
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply src/vault-mcp/search/search-queries.ts:386]: Fixed — added a length guard after `rerankPairs` returns. If the score count doesn't match the candidate count, `tryRerank` logs a warning and returns `null` (falls back to RRF-only ordering), preventing silent `NaN` propagation through `blendScores`.
